feat(StatusQ.Components): Adding StatusTagSelector component

alexandraB99 opened the pull request:
Added StatusTagSelector component needed for creating new chat channels, either one on one or group based on updated designs on Figma

Also added corresponding page in API Documentation

Closes #526

https://user-images.githubusercontent.com/31625338/150600131-9f9ed21b-b3a1-4efc-9637-131a727dc2cf.mov

https://user-images.githubusercontent.com/31625338/150442686-f24ff26e-03a4-47a3-a0fb-337e6a387768.mov

PascalPrecht commented on src/StatusQ/Controls/StatusTagSelector.qml:
```diff
@@ -0,0 +1,109 @@
+import StatusQ.Core 0.1
+import StatusQ.Core.Theme 0.1
+
+//TODO check colors, sizes, margins, general clean up
```
So.. is this component a work-in-progress?

alexandraB99 replied:
more or less. I need to double check colors, fonts, etc but I believe most of functionality is already there

PascalPrecht replied:
Gotcha!

PascalPrecht commented on src/StatusQ/Controls/StatusTagSelector.qml:
```diff
@@ -0,0 +1,109 @@
+                Layout.preferredWidth: 22
+                Layout.alignment: Qt.AlignVCenter | Qt.AlignLeft
+                color: Theme.palette.baseColor1
+                text: "To: "
```
This needs to be exposed for i18n

alexandraB99 replied:
done

PascalPrecht commented on src/StatusQ/Controls/StatusTagSelector.qml:
```diff
@@ -0,0 +1,109 @@
+                        anchors.left: nameText.right
+                        background: null
+                        contentItem: StatusIcon {
+                            color: Theme.palette.white
```
Does this work across themes?

alexandraB99 replied:
I see all status components are using this, is this not supposed to be working?

PascalPrecht replied:
Ah, my point was rather that, what is `white` here should probably not be `white` across all themes. Judging by other designs and components (e.g. StatusBadge), this should most likely become black in dark theme. So it has to be `Theme.palette.indirectColor`.

alexandraB99 replied:
done

PascalPrecht commented on src/StatusQ/Controls/StatusTagSelector.qml:
```diff
@@ -0,0 +1,109 @@
+                Layout.alignment: Qt.AlignVCenter | Qt.AlignRight
+                font.pixelSize: 10
+                color: Theme.palette.dangerColor1
+                text: "5 USER LIMIT REACHED"
```
This is a tag selector component that doesn't necessarily know the semantics of "users".

So few thoughts:

1. This message shouldn't be hardcoded (needs to be i18n'able anyways)
2. Would be cool if this component would leverage `StatusValidator` and then this message would be retrieved from the validator instance

Do you think it's feasible to replace `TextEdit` with `StatusBaseInput` (and tweak it as needed) so we get validation capabilities?

Ultimately, all (custom) controls should have validation built-in following the same APIs

alexandraB99 replied:
Thought about it (didn't try though), the problem is the internal Rectangle it has that paints the grey background and that's not customizable, unless I hide it, but still the red borders and the whole "invalid state" UI will look weird. Other than this, I don't see any invalid state [here](https://www.figma.com/file/IPpvkpDWabBKJTeo6bFop0/Kuba⎜Desktop?node-id=391%3A100984)

alexandraB99 replied:
done

PascalPrecht replied:
Okay, that's fair for now!

PascalPrecht reviewed (changes requested):
That's a cool start @alexandraB99 !

There are some things that need change.
Also:

1. Can you add a page component for this component accordingly, so it can actually be tested? I had no way to give this a spin...
2. Bonus points if the commit message includes the `Status.Controls` scope

Let me know what you think!

PascalPrecht commented:
Ah, seeing now that you flagged it as "don't merge" so I guess it **is** a WIP.

alexandraB99 commented:
> That's a cool start @alexandraB99 !
> 
> There are some things that need change. Also:
> 
> 1. Can you add a page component for this component accordingly, so it can actually be tested? I had no way to give this a spin...
> 2. Bonus points if the commit message includes the `StatusQ.Controls` scope
> 
> Let me know what you think!

@PascalPrecht yes, that is WIP :) Nevertheless I added the page and updated the commit message. It's good that I receive your comments from now so that I can work on those in one round :)

PascalPrecht commented on src/StatusQ/Controls/StatusTagSelector.qml:
```diff
@@ -0,0 +1,121 @@
+                    positionViewAtEnd();
+                }
+
+                delegate: Rectangle {
```
Realising now that this is the actual tag component.
I wonder if we should make some sort of button out of that. At, least, it should have proper hover effects/visuals (even if only on the "X" icon).

alexandraB99 replied:
this is the delegate, correct. There is no hovered state defined on Figma for this nevertheless, the `StatusIconTabButton` is used for the close icon which already has a hovered effect.

PascalPrecht commented on src/StatusQ/Controls/StatusTagSelector.qml:
```diff
@@ -0,0 +1,121 @@
+                    }
+                    ToolButton {
+                        width: 24
+                        height: 24
```
Not exactly sure what `ToolButton` is, but the size seems to be off. The icon needs to be rendered a bit bigger.

Also, can you ensure this turns into a cursor pointer when hovered?

alexandraB99 replied:
part of WIP, changed to `StatusIconTabButton`

PascalPrecht commented on src/StatusQ/Controls/StatusTagSelector.qml:
```diff
@@ -0,0 +1,121 @@
+                Layout.preferredWidth: 22
+                Layout.alignment: Qt.AlignVCenter | Qt.AlignLeft
+                color: Theme.palette.baseColor1
+                text: qsTr("To: ")
```
This still needs to be exposed. Probably as something like `toLabel`.

Remember that Status Desktop replaces `qsTr()` with `qsTrId()` in the sources (including vendors). That's why we expose all i18n'able labels so consumers of components can decide how they want to do i18n (e.g. using `qsTr()` or any other method)

alexandraB99 replied:
exposed :)

PascalPrecht commented on sandbox/demoapp/CreateChatView.qml:
```diff
@@ -0,0 +1,215 @@
+            onClicked: {
+                print("create chat...");
+            }
+        }
```
Cosmetics, but we can probably remove this one. The notification bell in the designs come from the `StatusChatToolBar` so I guess we 

1. Either use that here and draw the `StatusTagSelector` on top of it
2. Teach `StatusChatToolBar` to render components inside of it (like `StatusTagSelector`)
3. Just remove this notification thingy and don't assume that this is supposed to be the actual chat tool bar

Up to you!

alexandraB99 replied:
you are right, removed!

PascalPrecht reviewed (changes requested):
Thanks for the update @alexandraB99 !
This is getting in good shape. As you said, the base functionality is there and some visuals need to be changed.

I've left some more inline comments to address those.

In addition:

1. I think at the very least, users should be able to backspace items from the list. Being only able to "x" them out with the mouse is a big downside. 
2. The "X" of the tag component needs hover indicators. There aren't any in the design, so I'd check with the designers what that should look like (in both, light and dark mode)
3. Wonder if even clicking the entire tag should remove it (not just the "X"), in which case it needs updated hover visuals as well. But that is something design should decide.
4. There's this thing where the "too many users" message reserves a good amount of spaces that results in funny UI. In the screen shot below there was enough visual space, still the latest item pushed the tag list into the scroll space. 

<img width="603" alt="Screenshot 2022-01-21 at 11 21 12" src="https://user-images.githubusercontent.com/445106/150510452-b263300d-72bf-48c3-b21d-14510d9ea11c.png">

Maybe we can week the available space for tags as big as the input **until** the limit is reached?

5. Darkmode needs some love, in the component as well as in the demo application
<img width="1336" alt="Screenshot 2022-01-21 at 11 23 31" src="https://user-images.githubusercontent.com/445106/150510770-e3ae40fd-669b-4cc9-8d99-3d1a76399da5.png">

Great work so far!!

alexandraB99 commented on src/StatusQ/Controls/StatusTagSelector.qml:
```diff
@@ -0,0 +1,121 @@
+                        anchors.left: parent.left
+                        anchors.leftMargin: 8
+                        anchors.verticalCenter: parent.verticalCenter
+                        color: Theme.palette.white
```
done

alexandraB99 commented:
> Thanks for the update @alexandraB99 ! This is getting in good shape. As you said, the base functionality is there and some visuals need to be changed.
> 
> I've left some more inline comments to address those.
> 
> In addition:
> 
> 1. I think at the very least, users should be able to backspace items from the list. Being only able to "x" them out with the mouse is a big downside.
> 2. The "X" of the tag component needs hover indicators. There aren't any in the design, so I'd check with the designers what that should look like (in both, light and dark mode)
> 3. Wonder if even clicking the entire tag should remove it (not just the "X"), in which case it needs updated hover visuals as well. But that is something design should decide.
> 4. There's this thing where the "too many users" message reserves a good amount of space that results in funny UI. In the screen shot below there was enough visual space, still the latest item pushed the tag list into the scroll space.
> 
> <img alt="Screenshot 2022-01-21 at 11 21 12" width="603" src="https://user-images.githubusercontent.com/445106/150510452-b263300d-72bf-48c3-b21d-14510d9ea11c.png">
> 
> Maybe we can make the available space for tags as big as the input **until** the limit is reached?
> 
> 5. Darkmode needs some love, in the component as well as in the demo application
> 
> <img alt="Screenshot 2022-01-21 at 11 23 31" width="1336" src="https://user-images.githubusercontent.com/445106/150510770-e3ae40fd-669b-4cc9-8d99-3d1a76399da5.png">
> 
> Great work so far!!

2,4,5 fixed, 1 and 3 I would check with designers since nothing is specified in Figma.

PascalPrecht commented:
Thanks @alexandraB99 !
It seems the icon button used for the rendering the close icon still needs some tweaking. Below you can see a screenshot of both figma and sandbox. Notice how the "X" seems to be too big, and I believe the `leftMargin` needs to be smaller too.

In addition, notice how the "Tracy" tag doesn't have an "X"? The icon seems to disappear as soon as you hover the tag...

<img width="654" alt="Screenshot 2022-01-24 at 15 56 44" src="https://user-images.githubusercontent.com/445106/150806543-52c4ab8c-c0c7-4a7c-8f9e-d0a4467a4a66.png">

This breaks in both themes.

> 2,4,5 fixed, 1 and 3 I would check with designers since nothing is specified in Figma.

Sounds good!

alexandraB99 commented:
@PascalPrecht, I updated the spaces and sizes, regarding the color is the hover state default color from the `StatusIconTabButton` component, not sure what the hovered state of this component should be as it's not defined in Figma.

@John-44 button fixed. 
<img width="1336" alt="no_contact" src="https://user-images.githubusercontent.com/31625338/151838734-c6271107-4a46-4eb6-9ecf-f5a961540bb8.png">
<img width="1336" alt="contact" src="https://user-images.githubusercontent.com/31625338/151838744-238f40f9-f2e2-44ee-a5de-ad86d580da74.png">

A few questions:

- Should users be able to delete items from the list by pressing backspace?
- Wonder if even clicking the entire tag should remove it (not just the "X"), in which case it needs updated hover visuals as well. But that is something design should decide.???
- What is the hover state of the name tag delegate?

John-44 commented:
> A few questions:
> 
> * Should users be able to delete items from the list by pressing backspace?
> * Wonder if even clicking the entire tag should remove it (not just the "X"), in which case it needs updated hover visuals as well. But that is something design should decide.???
> * What is the hover state of the name tag delegate?

@alexandraB99 good questions! :-)

- Yes users should be able to delete items by pressing backspace (thanks for bringing this up)
- Re. clicking the entire tag should remove it: good idea, I don't see any reason not to do this because the tag has no other on-click action, and it would make touch interaction easier.  +1, good idea :-)

alexandraB99 commented:
> > A few questions:
> > 
> > * Should users be able to delete items from the list by pressing backspace?
> > * Wonder if even clicking the entire tag should remove it (not just the "X"), in which case it needs updated hover visuals as well. But that is something design should decide.???
> > * What is the hover state of the name tag delegate?
> 
> @alexandraB99 good questions! :-)
> 
> * Yes users should be able to delete items by pressing backspace (thanks for bringing this up)
> * Re. clicking the entire tag should remove it: good idea, I don't see any reason not to do this because the tag has no other on-click action, and it would make touch interaction easier.  +1, good idea :-)

thanks for the quick answer @John-44 , all these was @PascalPrecht's idea btw :) I have added both requirements, the only thing missing is the hovered state of the name tag, what would be that?

John-44 commented:
> the only thing missing is the hovered state of the name tag, what would be that?

@alexandraB99 For now let's go with something like this:
<img width="541" alt="Screenshot_2022-02-01_at_14 57 14" src="https://user-images.githubusercontent.com/22846356/151982183-73f5d0ea-a87a-47c3-970c-3a13db6e7e4f.png">

We will want to update this to something different later when the design system is more established (e.g. when David is at the point of defining all the button behaviours), but for now we can use this so development isn't blocked

alexandraB99 commented:
@John-44 thanks a lot, done
<img width="1336" alt="hovered" src="https://user-images.githubusercontent.com/31625338/152028811-2ccf814a-8274-4d1b-85a6-ffa8a29eaa3c.png">


Nevertheless, fyi, we shall wait for this PR https://github.com/status-im/StatusQ/pull/546 to get merged first before merging this one

PascalPrecht commented:
@John-44 If we can't get an opinion about this from design, I'd rather have us stay consistent and use the same behaviour (color-wise) we have in other buttons in our design system.

This button variant here:
<img width="100" alt="Screenshot 2022-02-01 at 20 50 39" src="https://user-images.githubusercontent.com/445106/152040754-6ac68e1c-a554-480b-b6b6-c29b54ecd295.png">

Looks like this in dark mode:
<img width="95" alt="Screenshot 2022-02-01 at 20 51 03" src="https://user-images.githubusercontent.com/445106/152040810-c8e5d01c-0cc2-49f1-b4dd-1d545705492b.png">

And the hovers look like this:
<img width="110" alt="Screenshot 2022-02-01 at 20 51 13" src="https://user-images.githubusercontent.com/445106/152040839-76c1656a-81b3-4752-b3c1-8bb432675cb3.png">
<img width="108" alt="Screenshot 2022-02-01 at 20 51 29" src="https://user-images.githubusercontent.com/445106/152040877-b761e8ed-f184-49aa-91cf-e1475e38225a.png">

So instead of introducing this "red" color which, honestly, doesn't go well with the blue, how about we only change the color of the entire button and just in the same way as the one above? That way it's consistent and we don't have this funny new thing :D 

Wdyt?

John-44 commented:
> And the hovers look like this: <img alt="Screenshot 2022-02-01 at 20 51 13" width="110" src="https://user-images.githubusercontent.com/445106/152040839-76c1656a-81b3-4752-b3c1-8bb432675cb3.png"> <img alt="Screenshot 2022-02-01 at 20 51 29" width="108" src="https://user-images.githubusercontent.com/445106/152040877-b761e8ed-f184-49aa-91cf-e1475e38225a.png">
> 
> So instead of introducing this "red" color which, honestly, doesn't go well with the blue, how about we only change the color of the entire button and just in the same way as the one above? That way it's consistent and we don't have this funny new thing :D
> 
> Wdyt?

@PascalPrecht yes what you are suggesting above is much better, +1 go with that :-)  

cc'ing @alexandraB99

alexandraB99 commented:
Ok @John-44 , @PascalPrecht I have changed the hovered to state as per @PascalPrecht's proposal. I have also moved StatusTagSelector to components as it makes more sense for it to be a component than a control. I also added the corresponding page in the API Documentation section.
<img width="1336" alt="normal" src="https://user-images.githubusercontent.com/31625338/152078034-4c63282c-916d-464e-9aa3-6cc9e7e00068.png">
<img width="1336" alt="hovered" src="https://user-images.githubusercontent.com/31625338/152078046-a8001278-69a1-416b-8b7f-c98c5e78ccd7.png">
<img width="1336" alt="tagselectorpage" src="https://user-images.githubusercontent.com/31625338/152078051-012ae6df-bf5e-479c-94cb-14657d6d54bf.png">

PascalPrecht commented:
Amazing job @alexandraB99 !
This looks very good now.

I've rebased it for you and landed it as https://github.com/status-im/StatusQ/commit/094dee492825520b47d3f4941a6720020dcbe6d3
